src/librbd/cache/pwl/InitRequest.cc

```diff
@@ -148,11 +150,48 @@ void InitRequest<I>::handle_set_feature_bit(int r) {
     lderr(cct) << "failed to set feature bit: " << cpp_strerror(r)
                << dendl;
     save_result(r);
-  } else if (m_image_ctx.discard_granularity_bytes) {
-    ldout(cct, 1) << "RWL image cache is enabled and "
-                  << "set discard_granularity_bytes = 0." << dendl;
-    m_image_ctx.discard_granularity_bytes = 0;
+
+    shutdown_image_cache();
+  } else {
+    if (m_image_ctx.discard_granularity_bytes) {
+      ldout(cct, 1) << "RWL image cache is enabled and "
+		    << "set discard_granularity_bytes = 0." << dendl;
+      m_image_ctx.discard_granularity_bytes = 0;
+    }
+
+    // Register RWL dispatch
+    auto image_dispatch = new cache::pwl::RWLImageDispatch<I>(&m_image_ctx, m_image_cache);
+
+    m_image_ctx.io_image_dispatcher->register_dispatch(image_dispatch);
+
+    finish();
   }
+}
+
+template <typename I>
+void InitRequest<I>::shutdown_image_cache() {
+  CephContext *cct = m_image_ctx.cct;
+  ldout(cct, 10) << dendl;
+
+  using klass = InitRequest<I>;
+  Context *ctx = create_context_callback<klass, &klass::handle_shutdown_image_cache>(
+    this);
+  m_image_cache->shut_down(ctx);
+}
+
+template <typename I>
+void InitRequest<I>::handle_shutdown_image_cache(int r) {
+  CephContext *cct = m_image_ctx.cct;
+  ldout(cct, 10) << dendl;
+
+  if (r < 0) {
+    lderr(cct) << "failed to close image cache: " << cpp_strerror(r)
+               << dendl;
+  } else {
```

Comment:
Nit: probably no need for the `else` clause -- you will have a memory leak if you don't free the cache

Reply:
Done.